File: leetcode/python/300-longest-increasing-subsequence.py
# Author: RT
# Date: 2022-07-17T11:01:19.180Z
# URL: https://leetcode.com/problems/longest-increasing-subsequence/


import bisect


class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        """Patience sort"""
        n = len(nums)
        tails = [nums[0]]

        for i in range(1, n):
            if nums[i] > tails[-1]:
                # extend the current longest LIS
                tails.append(nums[i])
            else:
                # exisiting length of LIS is found, replace end element with smaller value
                tails[bisect.bisect_left(tails, nums[i])] = nums[i]

        return len(tails)

    def lengthOfLIS__dp(self, nums: list[int]) -> int:
        n = len(nums)
        dp = [1 for _ in range(n)]
        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)

        return max(dp)

File: leetcode/rust/300-longest-increasing-subsequence.rs
// Author: RT
// Date: 2022-07-17T11:41:11.631Z
// URL: https://leetcode.com/problems/longest-increasing-subsequence/

impl Solution {
    pub fn length_of_lis(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut tails = vec![nums[0]];

        for i in 1..n {
            let curr = nums[i];
            if curr > tails[tails.len() - 1] {
                tails.push(curr);
            } else {
                // or use Vec.partition_point method
                let li = Self::bisect_left(&tails, curr);
                tails[li] = curr;
            }
        }

        tails.len() as i32
    }

    fn bisect_left(xs: &Vec<i32>, x: i32) -> usize {
        let (mut l, mut r) = (0 as usize, xs.len());
        while l < r {
            let mid = l + (r - l) / 2;
            if xs[mid] >= x {
                r -= 1;
            } else {
                l += 1;
            }
        }

        l
    }
}
